### PR TITLE
Embed translations and icons for mapeditor

### LINKED ISSUE #3522: Qt6.7 it breaks compilations on Linux
**Describe the bug**
After updating Qt 6.6 stable libraries to 6.7 beta, VCMI stopped compiling correctly.

**Logs**
```
 -- Configuring done (0.8s)
DEBUG util.py:446:  CMake Error at /usr/lib64/qt6/lib/cmake/Qt6LinguistTools/Qt6LinguistToolsMacros.cmake:483 (add_custom_command):
DEBUG util.py:446:    Attempt to add a custom rule to output
DEBUG util.py:446:      /builddir/build/BUILD/vcmi-1.4.3/build/czech.qm.rule
DEBUG util.py:446:    which already has a custom rule.
DEBUG util.py:446:  Call Stack (most recent call first):
DEBUG util.py:446:    /usr/lib64/qt6/lib/cmake/Qt6LinguistTools/Qt6LinguistToolsMacros.cmake:698 (qt6_add_lrelease)
DEBUG util.py:446:    /usr/lib64/qt6/lib/cmake/Qt6LinguistTools/Qt6LinguistToolsMacros.cmake:678:EVAL:1 (qt6_add_translations)
DEBUG util.py:446:    CMakeLists.txt:DEFERRED
DEBUG util.py:446:  CMake Error at /usr/lib64/qt6/lib/cmake/Qt6LinguistTools/Qt6LinguistToolsMacros.cmake:483 (add_custom_command):
DEBUG util.py:446:    Attempt to add a custom rule to output
DEBUG util.py:446:      /builddir/build/BUILD/vcmi-1.4.3/build/english.qm.rule
DEBUG util.py:446:    which already has a custom rule.
DEBUG util.py:446:  Call Stack (most recent call first):
DEBUG util.py:446:    /usr/lib64/qt6/lib/cmake/Qt6LinguistTools/Qt6LinguistToolsMacros.cmake:698 (qt6_add_lrelease)
DEBUG util.py:446:    /usr/lib64/qt6/lib/cmake/Qt6LinguistTools/Qt6LinguistToolsMacros.cmake:678:EVAL:1 (qt6_add_translations)
DEBUG util.py:446:    CMakeLists.txt:DEFERRED
```

full build log here: https://file-store.openmandriva.org/api/v1/file_stores/d5cdd5331436b553420db6122ab1e9dfc6276868.log?show=true

**To Reproduce**
1. Upgrade Qt libs to 6.7 beta
2. Start compilation VCMI 1.4.X
3. See build error

**Expected behavior**
All compile fine on Qt6.7

**Actual behavior**
Before upgrading to Qt6.7, game compiling fine, without single issue with previous Qt6 libs like 6.5 or 6.6. The problem only appeared after updating to 6.7.
I have this problem on OpenMandriva Cooker/ROME because my distro already deliver Qt6.7 libs.

I initially contacted about this issue on discord.

Either way, I suspect this issue will start affecting other systems as they migrate to Qt6.7.

**Version**
 - OS: Linux OpenMandriva Cooker/ROME
 - Version: 1.4.3

### PR DESCRIPTION
**Pull request by Gigas002**
Together with #3817 should fix #3522
I'm not experienced with c++, and just replicated the job done in d82f475b33150afbaf1f1bfcd1aa50592a519b12

### CODE REVIEW

File: mapeditor/CMakeLists.txt [hunk 1]
```diff
@@ -189,23 +219,9 @@ target_include_directories(vcmieditor
...
-# Install icons and desktop file on Linux
+
+# Install desktop file on Linux
 if(NOT WIN32 AND NOT APPLE)
 	install(FILES "vcmieditor.desktop" DESTINATION share/applications)
-	install(FILES "icons/mapeditor.32x32.png"   DESTINATION share/icons/hicolor/32x32/apps RENAME vcmieditor.png)
-	install(FILES "icons/mapeditor.48x48.png"   DESTINATION share/icons/hicolor/48x48/apps RENAME vcmieditor.png)
-	install(FILES "icons/mapeditor.64x64.png"   DESTINATION share/icons/hicolor/64x64/apps RENAME vcmieditor.png)
-	install(FILES "icons/mapeditor.128x128.png" DESTINATION share/icons/hicolor/128x128/apps RENAME vcmieditor.png)
-	install(FILES "icons/mapeditor.256x256.png" DESTINATION share/icons/hicolor/256x256/apps RENAME vcmieditor.png)
```
Comment (kambala-decapitator) [inline mapeditor/CMakeLists.txt:210]: looks like this part disappeared, please revert this change. You can also change it to a loop, see similar case here: https://github.com/vcmi/vcmi/pull/3668/files#diff-8404a76cc315289a505a0b9e571242cb4f5ba0c122fb86382329a4b714425ac3

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]
icons no longer appear in the editor. Seems you missed to add the via code after deleting from .ui file. Have you tested your changes? :)

Screenshot 2024-06-05 at 10 09 52

File: mapeditor/mainwindow.cpp [hunk 2]
```diff
@@ -206,6 +194,35 @@ MainWindow::MainWindow(QWidget* parent) :
...
 	ui->setupUi(this);
+
+	setWindowIcon(QIcon{":/icons/menu-game.png"});
+	ui->toolBrush->setIcon(QIcon{":/icons/brush-1.png"});
+	ui->toolBrush2->setIcon(QIcon{":/icons/brush-2.png"});
+	ui->toolBrush4->setIcon(QIcon{":/icons/brush-4.png"});
+	ui->toolLasso->setIcon(QIcon{":/icons/brush-3.png"});
+	ui->toolLine->setIcon(QIcon{":/icons/brush-7.png"});
+	ui->toolArea->setIcon(QIcon{":/icons/brush-5.png"});
+	ui->toolFill->setIcon(QIcon{":/icons/brush-6.png"});
```
Comment (kambala-decapitator) [inline mapeditor/mainwindow.cpp:205]: these icons deserve a better name :)

Comment (Gigas002) [reply mapeditor/mainwindow.cpp:205]: I agree! :) Should I rename the original files accordingly to a ui element these are used at?

**Comment by Gigas002** [comment]
> do you have lrelease on your machine? 

Yes I do. From the description it provides I suppose it's part of `qt5-tools`

**Comment by kambala-decapitator** [comment]
> > do you have lrelease on your machine?
> 
> Yes I do. From the description it provides I suppose it's part of `qt5-tools`

on CI we install qtbase5-dev and qttools5-dev, that's probably the difference

File: mapeditor/mainwindow.cpp (hunk 2 above)
Comment (kambala-decapitator) [reply mapeditor/mainwindow.cpp:205]: for sure. Icon name should relate to icon content.

**Comment by Gigas002** [comment]
Alright, I think I found out the way to fix this. Using the full path for qt5 translation generation shouldn't break anything?

```diff
- set_source_files_properties(${launcher_TS} PROPERTIES OUTPUT_LOCATION translation)
+ set_source_files_properties(${launcher_TS} PROPERTIES OUTPUT_LOCATION "${CMAKE_CURRENT_BINARY_DIR}/${translationsDir}")
```

**Comment by kambala-decapitator** [comment]
> Alright, I think I found out the way to fix this. Using the full path for qt5 translation generation shouldn't break anything?
> 
> ```diff
> - set_source_files_properties(${launcher_TS} PROPERTIES OUTPUT_LOCATION translation)
> + set_source_files_properties(${launcher_TS} PROPERTIES OUTPUT_LOCATION "${CMAKE_CURRENT_BINARY_DIR}/${translationsDir}")
> ```

yes, seems fine. Will add the same to the android PR then.

maybe behavior is cmake version dependent? what's your version btw?

**Comment by Gigas002** [comment]
I'm using version `3.29.4`.
The PR is ready for re-review, I think, could you please take a look?

**Review by kambala-decapitator — APPROVED** [review]
> I'm using version 3.29.4.

that's fine. No idea why such difference in behavior then. But good that you found the solution!

**Comment by kambala-decapitator** [comment]
> Alright, I think I found out the way to fix this. Using the full path for qt5 translation generation shouldn't break anything?
> 
> ```diff
> - set_source_files_properties(${launcher_TS} PROPERTIES OUTPUT_LOCATION translation)
> + set_source_files_properties(${launcher_TS} PROPERTIES OUTPUT_LOCATION "${CMAKE_CURRENT_BINARY_DIR}/${translationsDir}")
> ```

apparently that was a bad decision. Now file `launcher/translations.qrc` in the build dir has absolute paths which is not how it should be. So this is not a solution. While it probably works for local install on desktop systems, relocating would break paths.

File: mapeditor/CMakeLists.txt [hunk 3]
```diff
@@ -107,24 +107,59 @@ set(editor_FORMS
...
+	"${translationsDir}/polish.ts"
+	"${translationsDir}/portuguese.ts"
+	"${translationsDir}/russian.ts"
+	"${translationsDir}/spanish.ts"
+	"${translationsDir}/ukrainian.ts"
+	"${translationsDir}/vietnamese.ts"
 )
+if(ENABLE_TRANSLATIONS)
+	if(TARGET Qt5::Core)
+		set_source_files_properties(${editor_TS} PROPERTIES OUTPUT_LOCATION "${CMAKE_CURRENT_BINARY_DIR}/${translationsDir}")
```
Comment (kambala-decapitator) [inline mapeditor/CMakeLists.txt:130]: must be relative path

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

**Comment by Gigas002** [comment]
I see. This is indeed incorrect behavior.
How about we create a directory first and leave the second line unchanged? It seems like it fails to create directory and because of that translation files are not copied inside.

```diff
- set_source_files_properties(${editor_TS} PROPERTIES OUTPUT_LOCATION "${CMAKE_CURRENT_BINARY_DIR}/${translationsDir}")
+ file(MAKE_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/${translationsDir}")
+ set_source_files_properties(${editor_TS} PROPERTIES OUTPUT_LOCATION "${translationsDir}")
```

The `translation.qrc` will have relative paths in that case:

```
<!DOCTYPE RCC>
<RCC version="1.0">
<qresource prefix="/">
<file>translation/chinese.qm</file>
<file>translation/czech.qm</file>
<file>translation/english.qm</file>
<file>translation/french.qm</file>
<file>translation/german.qm</file>
<file>translation/polish.qm</file>
<file>translation/portuguese.qm</file>
<file>translation/russian.qm</file>
<file>translation/spanish.qm</file>
<file>translation/ukrainian.qm</file>
<file>translation/vietnamese.qm</file>

</qresource>
</RCC>
```

**Comment by kambala-decapitator** [comment]
that should work I believe, let's try

**Review by kambala-decapitator — APPROVED** [review]

**Comment by Gigas002** [comment]
So... I've seen the Ivan's response in the discord a while ago, that this was merged to a wrong branch, so it's not in 1.5.3. What does that mean for this PR? Will it be included in next release?

**Comment by kambala-decapitator** [comment]
> So... I've seen the Ivan's response in the discord a while ago, that this was merged to a wrong branch, so it's not in 1.5.3. What does that mean for this PR? Will it be included in next release?

Please cherry-pick your commits to the **beta** branch
